### DISCUSSION TITLE
how reset/reinstance a componenent

### DISCUSSION THREAD

**Discussion by ti-integra** [slint-ui/slint#5688] [Q&A] [answered]
in the demo gallery when change sidebar the widget reset,

in [my code](https://github.com/tredeneo/simple-inventary-management/blob/main/ui/main.slint) when I change de siderbar the widget/component is in the last state and not a new

example in another lib, iced sidebar example

currently occur (in slint its the same result)

```rust
    fn update(&mut self, message: Message) {
        match message {
            Message::TabSelected(selected) => self.active_tab = selected,
            Message::Login(message) => self.login_tab.update(message),
            Message::Ferris(message) => self.ferris_tab.update(message),
            Message::Counter(message) => self.counter_tab.update(message),
            Message::Settings(message) => self.settings_tab.update(message),
            Message::TabClosed(id) => println!("Tab {:?} event hit", id),
        }
    }

```
 what I want (this is the result that I want code in slint)

```rust
    fn update(&mut self, message: Message) {
        match message {
            Message::TabSelected(selected) => {
                self.active_tab = selected.clone();
                self.reset_tab(selected);
            }
            Message::Login(message) => self.login_tab.update(message),
            Message::Ferris(message) => self.ferris_tab.update(message),
            Message::Counter(message) => self.counter_tab.update(message),
            Message::Settings(message) => self.settings_tab.update(message),
            Message::TabClosed(id) => println!("Tab {:?} event hit", id),
        }
    }
    fn reset_tab(&mut self,tab_id:TabId){
        
        match tab_id {
            TabId::Login => self.login_tab = LoginTab::new(),
            TabId::Ferris => self.ferris_tab = FerrisTab::new(),
            TabId::Counter => self.counter_tab = CounterTab::new(),
            TabId::Settings => self.settings_tab = SettingsTab::new(),
        }
    }
```
anyone know how I can reach the second example result works in slint?  or what's wrong with my code

[my code](https://github.com/tredeneo/simple-inventary-management/tree/main/ui)

**Comment by hunger** [comment]
This is a simplified version of your code:

```slint
export component App inherits Window {
    HorizontalLayout {
        side-bar := SideBar {
            title: @tr("Slint Widgets Gallery");
            model: [ "1", "2", "3" ];
        }

        if(side-bar.current-item == 0): LineEdit { text: "zero"; }
        if(side-bar.current-item == 1): LineEdit { text: "one"; }
        if(side-bar.current-item == 2): LineEdit { text: "two"; }
    }
}
```

The `if` makes sure only the selected widget exists. So when you switch, the previous widget gets deleted and the current one constructed.

Try this instead:

```slint
export component App inherits Window {
    HorizontalLayout {
        side-bar := SideBar {
            title: @tr("Slint Widgets Gallery");
            model: [ "1", "2", "3" ];
        }

        Rectangle {
            LineEdit { text: "zero"; visible: side-bar.current-item == 0; }
            LineEdit { text: "one"; visible: side-bar.current-item == 1; }
            LineEdit { text: "two"; visible: side-bar.current-item == 2; }
        }
    }
}
```

The Rectangle is there so all the `LineEdit` show up in the same area of the layout.

The widgets all exist at all times here, but only one of them is visible at a time. So they keep state...

I hope this help, and sorry for the delay in getting back to you!

**Reply by ti-integra** [reply]
```The if makes sure only the selected widget exists. So when you switch, the previous widget gets deleted and the current one constructed.```

my problem is this, I want that the previous deleted e constructed a new on current. but in my code occur the opposite, the previous is not been deleted

this happens in the side-bar, but for some reason this behavior does not occur in my code

**Reply by hunger** [reply]
Well, technically the `if` hides the widgets where the condition is false (they stay in the tree though), but it reinitialized the widget -- at least it did when I ran this in Slintpad :-)

I am not sure what you observe and what you would like to happen. You say you see the ecpected behavior in the sidebar, but not in your code. Which part of your code is that?

**Reply by ti-integra** [reply]
i think its in rust code, when the widget is initialized has this [code](https://github.com/tredeneo/simple-inventary-management/blob/main/src/ui/brand.rs) (one screen example ), but this is has only full execution in the first time, and every time that I update database I have to use ui_update() explicit, in my example in iced, every time that use screen::new() all data is reset (and I put this in the side-bar, so every time that I change screen, the screen is reseted)

in slint for every update that I have in database, I have to update explicit the screen

**Reply by hunger** [reply, marked as answer]
In slint you will need to update the model holding the data. The screen should then update automatically. At least that is the theory :-)

You seem to unconditionally update *everything* in your model. That will destroy and recreate the ui elements associated with those items in the model -- which will reset them.

Try to update only those rows that changed instead.

**Reply by ti-integra** [reply]
I thought that other option will be available, will follow with that you explained
